apm-agent-plugins/apm-log-shader-plugin/apm-log-shader-plugin-common/src/main/java/co/elastic/apm/agent/log/shader/AbstractLogShadingHelper.java

```diff
@@ -0,0 +1,146 @@
+        if (isShadingAppender(originalAppender)) {
+            return null;
+        }
+
+        Object shadeAppender = appenderToShadeAppender.get(originalAppender);
+        if (shadeAppender == null) {
+            synchronized (appenderToShadeAppender) {
+                if (!appenderToShadeAppender.containsKey(originalAppender)) {
+                    A createdAppender = createAndConfigureAppender(originalAppender, ECS_SHADE_APPENDER_NAME);
+                    appenderToShadeAppender.put(originalAppender, createdAppender != null ? createdAppender : NULL_APPENDER);
+                }
+            }
+            shadeAppender = appenderToShadeAppender.get(originalAppender);
+        }
+        return shadeAppender != NULL_APPENDER ? (A) shadeAppender : null;
+    }
+
+    public void stopShading(A originalAppender) {
+        synchronized (appenderToShadeAppender) {
+            Object shadeAppender = appenderToShadeAppender.remove(originalAppender);
+            if (shadeAppender != null) {
+                closeShadeAppender((A) shadeAppender);
+            }
+        }
+    }
+
+    /**
+     * Checks whether we should skip {@code append()} invocations for log events for the given appender.
+     * Log event appends should be skipped if they are replaced by ECS-formatted events, meaning if:
+     *  - shading is enabled by configuration AND
+     *  - replace is enabled by configuration AND
+     *  - there is a shade appender for this appender (there isn't when this appender IS A shade appender or it is not a file appender)
+     * @param appender the appender
+     * @return true if log events should be ignored for the given appender; false otherwise
+     */
+    public boolean shouldSkipAppend(A appender) {
+        return loggingConfiguration.getLogEcsReformatting() == LogEcsReformatting.REPLACE && getOrCreateShadeAppenderFor(appender) != null;
+    }
+
+    public boolean isShadingEnabled() {
+        LogEcsReformatting logEcsReformatting = loggingConfiguration.getLogEcsReformatting();
+        return logEcsReformatting == LogEcsReformatting.SHADE || logEcsReformatting== LogEcsReformatting.REPLACE;
+    }
+
+    /**
+     * Checks whether the given appender is a shading appender, so to avoid recursive shading
+     *
+     * @return true if the provide appender is a shading appender; false otherwise
+     */
+    private boolean isShadingAppender(A appender) {
+        //noinspection StringEquality
+        return getAppenderName(appender) == ECS_SHADE_APPENDER_NAME;
+    }
+
+    protected abstract String getAppenderName(A appender);
+
+    @Nullable
+    protected abstract A createAndConfigureAppender(A originalAppender, String appenderName);
+
+    @Nullable
+    protected String getServiceName() {
+        String serviceName = configuredServiceName;
+        AbstractSpan<?> active = tracer.getActive();
+        if (active != null) {
+            String runtimeServiceName = active.getTraceContext().getServiceName();
```

Comment:
This may yield different service names for the same webapp depending on whether the log has been created inside or outside of an active transaction.
Not sure what the alternative is, though.

Reply:
Yes, I am aware and agree this is not ideal. However, don't forget that when logging correlation is enabled, we also add `trace.id` and `transaction.id`. What I tried to do is keep these Log events in sync with the APM events that we report in that context. So I guess the question is: what's more important - the sync with APM events or the ability to aggregate for a single service name in the logging app?

Reply:
(I honestly don't have an answer, I think you are in a better position to judge, so we'll go with what you think is better)

Reply:
One thing I missed is that `getServiceName` only gets called in the context of `co.elastic.apm.agent.log.shader.AbstractLogShadingHelper#createAndConfigureAppender` which is only called once, when the first log after instrumenting the appender is logged.

This may or may not be within a transaction. Likely, it happens before the service name can be determined for a `.war` web application or before being able to detect the `spring.application.name`.

It's tricky. But at least, the service name should stay consistent throughout the lifetime of the application.

Reply:
That is correct, and now I recall we already had this discussion 🙂 
If we think it's important enough to maintain consistency (for future purposes as well), we can add the service name to the MDC and in the ECS logger first look for such key and only if it is not there, use the one configured during the appender configuration.
Really, it's your call (sorry, I didn't understand from the last comment if made this decision)

Reply:
I think it's important that the service name stays consistent regardless of whether or not a log is emitted during a transaction. So setting it once when creating the ecs formatter/encoder does make sense. But I think we can refine the way we gather the service name. See also https://github.com/elastic/apm-agent-java/pull/1261/files#r587523067